### PR TITLE
Run `calicoctl node` without launching docker

### LINKED ISSUE #564: It should be possible to set up a node without starting calico (felix/bird/confd)
Operators may want to choose their own way of deploying the `calico-node` components i.e. felix, bird and confd. We currently only support using a docker container, but we do have native packages in our openstack repos. In addition to running the components natively, operators may want to use other container formats such as rkt.

The expected way that we'll support this is to add a new `calicoctl node --runtime=<RUNTIME>` command. It can take the following values
* docker - which is the current and default behaviour
* none - which doesn't run felix/bird/confd at all.

It future we may add other options.

### PR DESCRIPTION
**Pull request by djosborne**
Base implementation to fix #564 . Adds the `--runtime` option with possible values of `docker` or `none`. This PR also patches some bugs encountered by `checksystem` and `status` which both assume a connection to the docker-daemon is available. 

Additional design is needed to provide more robust checksystem and status for dockerless systems, but I propose we raise separate issues in calico-docker to address improving these.

### CODE REVIEW

**Comment by robbrockbank** [comment]
I know this isn't up for review yet - but just a note:  Don't forget to also update the calicoctl documentation.  Ta!

File: calico_containers/calico_ctl/node.py [hunk 1]
```diff
@@ -35,6 +36,8 @@
                             "false" to run in the foreground.  When using
                             libnetwork, this may not be set to "false".
                             [default: true]
+  --runtime=<RUNTIME>       Specify how calico-node processes should be
```
Comment (spikecurtis) [inline calico_containers/calico_ctl/node.py]: "Specify how Calico services..."

I don't want to ossify around `calico-node` terminology.  Let's leave that term just to mean the Docker container image.

File: calico_containers/calico_ctl/node.py [hunk 2]
```diff
@@ -35,6 +36,8 @@
...
+                            launched.  Can be: docker, none. [default: docker]
```
Comment (spikecurtis) [inline calico_containers/calico_ctl/node.py]: We probably need some explanation of what "none" means.

File: calico_containers/calico_ctl/status.py [hunk 3]
```diff
@@ -35,6 +36,13 @@ def status(arguments):
     this file's docstring with docopt
     :return: None
     """
+    try:
+        docker_client.version()
```
Comment (spikecurtis) [inline calico_containers/calico_ctl/status.py]: Why not just wrap the call below in try except?

File: calico_containers/calico_ctl/node.py [hunk 4]
```diff
@@ -536,24 +543,31 @@ def warn_if_hostname_conflict(ip):
...
-                print_paragraph("WARNING: Hostname '%s' is already in use "
-                                "with IP address %s. Calico requires each compute host to "
-                                "have a unique hostname. If this is your first time "
-                                "running 'calicoctl node' on this host, ensure that "
-                                "another host is not already using the "
-                                "same hostname." % (hostname, ip))
+    try:
+        if docker_client.containers(filters={'name': 'calico-node'}):
+            return
+    except IOError:
```
Comment (spikecurtis) [inline calico_containers/calico_ctl/node.py]: I think we still want to do the hostname conflict warning, even if the user doesn't have Docker running.

File: calico_containers/calico_ctl/status.py [hunk 5]
```diff
@@ -35,6 +36,13 @@ def status(arguments):
...
+    except ConnectionError:
```
Comment (spikecurtis) [inline calico_containers/calico_ctl/status.py]: In this check you seem to be using "Is the Docker daemon running?" as a proxy for "Did the user start the node with `--runtime=none`?"  These aren't the same thing.  You could use `--runtime=none` even though Docker is running, for example. In that case, won't this function fail?

File: docs/calicoctl/node.md [hunk 6]
```diff
@@ -112,6 +119,10 @@ When running the `calicoctl node` command with the `--libnetwork` plugin, the
 command starts a container using the `calico/node-libnetwork` Docker image in 
 addition to starting the `calico/node` Docker image.
 
+The `--runtime` flag can be used to prevent calico from launching the
+calico-node container, instead allowing you to run the core processes
+on baremetal.
```
Comment (spikecurtis) [inline docs/calicoctl/node.md]: Bare metal is a bit misleading.  You might use this in a VM, for example.

"... run the core Calico services via some other means. For example, you might install directly on the host and execute them via a systemd unit."

File: docs/calicoctl/node.md [hunk 7]
```diff
@@ -112,6 +119,10 @@ When running the `calicoctl node` command with the `--libnetwork` plugin, the
 command starts a container using the `calico/node-libnetwork` Docker image in 
 addition to starting the `calico/node` Docker image.
 
+The `--runtime` flag can be used to prevent calico from launching the
+calico-node container, instead allowing you to run the core processes
```
Comment (spikecurtis) [inline docs/calicoctl/node.md]: "calico-node Docker container"

File: docs/calicoctl/node.md [hunk 8]
```diff
@@ -112,6 +119,10 @@ When running the `calicoctl node` command with the `--libnetwork` plugin, the
 command starts a container using the `calico/node-libnetwork` Docker image in 
 addition to starting the `calico/node` Docker image.
 
+The `--runtime` flag can be used to prevent calico from launching the
```
Comment (spikecurtis) [inline docs/calicoctl/node.md]: Calico

Comment (spikecurtis) [inline docs/calicoctl/node.md]: `--runtime=none` setting

File: docs/calicoctl/node.md [hunk 9]
```diff
@@ -71,7 +74,9 @@ Options:
 
 ### calicoctl node 
 
-This command starts a container using the `calico/node` Docker image. 
+This command performs two actions:
+1. Initialize this host for calico by setting first-time config in etcd.
+2. Start the `calico/node` Docker container.
```
Comment (spikecurtis) [inline docs/calicoctl/node.md]: (Optional) Start Calico services.  By default (or via `--runtime=docker`) the command does this by downloading the `calico/node` Docker image and running it in a container.

File: docs/calicoctl/node.md [hunk 10]
```diff
@@ -71,7 +74,9 @@ Options:
 
 ### calicoctl node 
 
-This command starts a container using the `calico/node` Docker image. 
+This command performs two actions:
+1. Initialize this host for calico by setting first-time config in etcd.
```
Comment (spikecurtis) [inline docs/calicoctl/node.md]: Calico

File: calico_containers/calico_ctl/status.py (hunk 3 above)
Comment (djosborne) [inline calico_containers/calico_ctl/status.py]: There's a lot of docker calls throughout `status` which gather info. Wrapping the whole thing in a try/except means that If one of the initial calls passes fine but the later ones do not, the "Docker is not running" message wouldn't accurately capture the real problem.

File: calico_containers/calico_ctl/node.py [hunk 11]
```diff
@@ -536,24 +548,33 @@ def warn_if_hostname_conflict(ip):
...
+                    # Calico-node isn't running on this host.
+                    # There may be another host using this hostname.
+                    print_paragraph("WARNING: Hostname '%s' is already in use "
+                            "with IP address %s. Calico requires each compute host to "
+                            "have a unique hostname. If this is your first time "
+                            "running 'calicoctl node' on this host, ensure that "
+                            "another host is not already using the "
+                            "same hostname." % (hostname, ip))
+            except IOError:
+                # Couldn't connect to docker to confirm calico-node is running.
```
Comment (spikecurtis) [inline calico_containers/calico_ctl/node.py]: I don't think we want to drop a warning if they are using `--runtime=none` here, or at least not this warning.

Comment (djosborne) [inline calico_containers/calico_ctl/node.py]: Ideally we'd detect if calico-felix is running on the system here, and if it is not, then drop the warning. Instead of designing a way to detect if Felix is running or not, I figured it'd be safer to always warn for now.

File: calico_containers/calico_ctl/status.py [hunk 12]
```diff
@@ -35,29 +36,34 @@ def status(arguments):
...
+
+            if result is not None:
+                print "Running felix version %s" % result.group(1)
+
+            print "\nIPv4 BGP status"
+            pprint_bird_protocols(4)
+            print "IPv6 BGP status"
+            pprint_bird_protocols(6)
+    except ConnectionError:
+        print "Docker is not running"
```
Comment (spikecurtis) [inline calico_containers/calico_ctl/status.py]: Can we make this message meaningful to a user that doesn't want to use Docker?  Ditto with other messages.  The messages are factually true, but with multiple runtimes it isn't clear what the user should expect to see.

I'm wondering if we should write some data to an OS file to make `calicoctl` stateful and more helpful.  It would help to know answers to questions like

 - has an admin run `calicoctl node` on this host before?
 - what runtime did they use the last time they ran it?

We could even log all the `calicoctl` commands, which would be super helpful for debugging issues in the field.

Comment (djosborne) [inline calico_containers/calico_ctl/status.py]: While I do agree with these suggestions, I don't believe they fall within the scope of the issue this PR is trying to address. My goal in this PR is only to allow users to launch a dockerless node. Improvements to `checksystem` and `statuts` to provide useful information when docker has launched in this way are definitely the next step, but require significant thought into how calicoctl is structured. I think it'd be best if we start that work in a new PR after a bit of designing

Comment (spikecurtis) [inline calico_containers/calico_ctl/status.py]: Sounds good.  Raise an issue so we don't lose track of it.

File: calico_containers/calico_ctl/node.py (hunk 11 above)
Comment (spikecurtis) [inline calico_containers/calico_ctl/node.py]: OK.  I'd forgotten that this only drops if they re-run and change the IP address, so it's not like we'll drop it every time if they put it in a startup script or whatever.

File: calico_containers/tests/unit/node_test.py [hunk 13]
```diff
@@ -166,6 +166,98 @@ def test_validate_arguments(self, case, sys_exit_called):
...
     @patch('calico_ctl.node._find_or_pull_node_image', autospec=True)
     @patch('calico_ctl.node._attach_and_stream', autospec=True)
+    def test_node_dockerless_start(self, m_attach_and_stream,
+                                   m_find_or_pull_node_image, m_docker,
+                                   m_docker_client, m_client, m_install_plugin,
+                                   m_warn_if_hostname_conflict, m_warn_if_unknown_ip,
+                                   m_get_host_ips, m_setup_ip, m_check_system,
+                                   m_os_makedirs, m_os_path_exists):
+        """
+        Test that the node_start function behaves as expected when
```
Comment (spikecurtis) [inline calico_containers/tests/unit/node_test.py]: "behaves as expected" is pretty meaningless in this context.  How about "node_start function does not make Docker calls when runtime=None"

**Comment by spikecurtis** [comment]
One minor nit, otherwise LGTM.

File: calico_containers/calico_ctl/node.py [hunk 14]
```diff
@@ -35,6 +36,11 @@
                             "false" to run in the foreground.  When using
                             libnetwork, this may not be set to "false".
                             [default: true]
+  --runtime=<RUNTIME>       Specify how calico services should be
```
Comment (tomdee) [inline calico_containers/calico_ctl/node.py]: Calico?

File: calico_containers/calico_ctl/node.py [hunk 15]
```diff
@@ -35,6 +36,11 @@
...
+                            launched.  When set to "docker", services will be
+                            launched via the calico-node container, whereas a
+                            value of "none" will not launch them
```
Comment (tomdee) [inline calico_containers/calico_ctl/node.py]: funny line break

File: calico_containers/calico_ctl/node.py [hunk 16]
```diff
@@ -536,24 +548,33 @@ def warn_if_hostname_conflict(ip):
...
+                            "same hostname." % (hostname, ip))
+            except IOError:
+                # Couldn't connect to docker to confirm calico-node is running.
                 print_paragraph("WARNING: Hostname '%s' is already in use "
-                                "with IP address %s. Calico requires each compute host to "
-                                "have a unique hostname. If this is your first time "
-                                "running 'calicoctl node' on this host, ensure that "
-                                "another host is not already using the "
-                                "same hostname." % (hostname, ip))
+                            "with IP address %s. Calico requires each compute host to "
```
Comment (tomdee) [inline calico_containers/calico_ctl/node.py]: Why did you change the alignment here?

File: calico_containers/calico_ctl/node.py (hunk 15 above)
Comment (djosborne) [inline calico_containers/calico_ctl/node.py]: I didn't want [default:all] on its own line, so I broke early....

File: calico_containers/calico_ctl/node.py (hunk 16 above)
Comment (tomdee) [inline calico_containers/calico_ctl/node.py]: You should stick this big text block in a variable.

**Comment by tomdee** [comment]
A couple of little tidyups but otherwise LGTM
